graph-http/src/intoresponse.rs

```diff
@@ -286,9 +316,9 @@ impl<'a> IntoResponseBlocking<'a, BlockingDownload> {
 // Async Impl
 
 impl<'a, T> IntoResponseAsync<'a, T> {
-    pub async fn json<U>(self) -> GraphResult<GraphResponse<U>>
+    pub async fn json<U, V>(self) -> GraphResult<GraphResponse<U>>
```

Comment:
It might be better to just use a struct that is a placeholder for T/phantom data.

This way you could have specific responses only for T if T is ODataNextLink

```rust
impl<'a, NextLink<T>> IntoResponseAsync<'a, NextLink<T>> {
    pub async fn json<T>>(self) -> **GraphResult<GraphResponse<U>>**
```

Something like that.

Reply:
I think I understand what you mean. Not sure how to do that but I can give it a go.


Would this not result in a "multiple definitions" for json() ?

Reply:
I tried to implement that as 
``` rust 
#[derive(Deserialize)]
pub struct NextLink<'a, V>
{
    value: Option<&'a mut Vec<V>>,
    next_link: Option<String>,
}

impl<V> NextLink<'_, V> {
    fn next_link(&self) -> Option<String> {
        self["@odata.nextLink"].as_str().map(|s| s.to_string())
    }

    fn value(&mut self) -> Option<&mut Vec<V>> {
        self["@odata.value"].as_array_mut()
    }
}

impl<'a, T> IntoResponseAsync<'a, NextLink<'_, T>> {
pub async fn json<U>(self) -> GraphResult<GraphResponse<U>> {

```
and the :crab: is not happy :
```
error[E0592]: duplicate definitions with name `json`
   --> graph-http\src\intoresponse.rs:334:1
    |
334 |   pub async fn json<U>(self) -> GraphResult<GraphResponse<U>>
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ duplicate definitions for `json`
...
376 | /     pub async fn json<U>(self) -> GraphResult<GraphResponse<U>>
377 | |         where
378 | |                 for<'de> U: serde::Deserialize<'de>,
    | |____________________________________________________- other definition for `json`
```

Am I missing something ?

Reply:
I was meaning more like how NoContent is used. But I gave the wrong example before

https://github.com/sreeise/graph-rs/blob/248f9298551e721042b7ee9a27e99ede3e5f0cce/graph-http/src/types/content.rs#L2

https://github.com/sreeise/graph-rs/blob/248f9298551e721042b7ee9a27e99ede3e5f0cce/graph-http/src/intoresponse.rs#L244

Its a struct that is basically a placeholder so that only request methods that return NoContent can use that specific implementation.

I was thinking with your method with_next_links (or whatever its called) you would return IntoResponse but with a different type. The method would still be part of the generic impl of IntoResponse but then it would return a more concrete type of IntoResponse:

```rust
#[derive(Clone, Debug, Default)]
pub struct NextLink;


impl<'a, T> IntoResponseBlocking<'a, T> {
    pub fn with_next_links(self) -> IntoResponseBlocking<'a, NextLink> {
        self
    }
}

impl <'a> IntoResponseBlocking<'a, NextLink> {
    pub fn json<U>(self) -> GraphResult<GraphResponse<U>>
        where
                for<'de> U: serde::Deserialize<'de>,
    {
         // ....
    }
}
```

This should prevent any issues with duplicate definitions since the type of T is concrete and has its own implementation where you can have a separate json method.

Reply:
Based on the above you could possibly try

```rust
impl<'a, T> IntoResponseBlocking<'a, T> {
    pub fn with_next_links(self) -> IntoResponseBlocking<'a, NextLink<'a, T>> {
        IntoResponse::new(self.client)
    }
}

impl<'a, T> IntoResponseBlocking<'a, NextLink<'a, T>> {
    pub fn send(self) -> GraphResult<GraphResponse<T>>
        where
                for<'de> T: serde::Deserialize<'de> + ODataNextLink,
    {
     /// ...
    }
}
```

And when calling its something like:

```rust
client.v1()
  .users()
  .list_user()
  .with_next_links::<serde_json::Value>()
  .send()
  .unwrap();

```

Reply:
It does indeed work when doing it like that.
Any reason that json() is so generic and not restricted to `T: serde::Deserialize` ?
Neither my work project nor tests fail when I move the `json()` to the 
```rust
impl<'a, T> IntoResponseBlocking<'a, T>
where
    for<'de> T: serde::Deserialize<'de>,
{
}
``` 
block.

We have 3 possibilities :
- name it `send`, but the naming does not follow the rest of the library
- name it `json`, but implies to restrict the generic method
- name it something like `json_with_next_links` and remove the `with_next_links_calls` method

Reply:
Nope, no specific reason for that outer T being generic except for the other two methods there, text and bytes, which dont need to implement Deserialize.

So you could move the json method to another implementation block with the T restricted to Deserialize and just use that. The text and bytes methods can just stay in the same place they are now with that generic T.